File: Python/JCM.py
import random

dog_list = ["Fido", "Spot", "Sparky"]

def dog_position():
    position = random.randint(0, 99)  # Generate a random position
    return position  # Return the generated position
  
def dog_walking(dog_location):
  time=random.randint(1,10)
  speed=random.randint(1,10)
  dog_new_location= dog_location+(time*speed)
  print("The dog is walking during " + str(time) + " seconds with a speed of "+ str(speed) +" m/s")
  return dog_new_location

def guess_new_position(dog_walking):
  guess=int(input("Guess the new position of the dog: "))
  return guess

def dog_guess(dog_location, guess):
  if guess==dog_location:
    print("Correct!")
  else:
    print("Incorrect!")
    print ("New location is " + str(dog_new_location))

for dog_name in dog_list:
  dog_location = dog_position()  # Get the location using the dog_position function
  print("Actual " + dog_name + " location is " + str(dog_location))
  dog_new_location = dog_walking(dog_location)
  guessing=guess_new_position(dog_new_location)
  dog_guess(dog_new_location, guessing)

File: C/JCM.c
#include <windows.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BCRYPT_USE_SYSTEM_PREFERRED_RNG 0x00000002
/* typedef minimal (évite d'inclure bcrypt.h) */
typedef LONG (WINAPI *BCryptGenRandom_t)(void *hAlgorithm, unsigned char *pbBuffer, unsigned long cbBuffer, unsigned long dwFlags);

void init_rng_with_bcrypt(void) {
    HMODULE h = LoadLibraryA("bcrypt.dll");
    if (!h) return; /* bcrypt.dll absent -> on peut retomber sur srand(time(NULL)) si besoin */

    BCryptGenRandom_t p = (BCryptGenRandom_t)GetProcAddress(h, "BCryptGenRandom");
    if (!p) { FreeLibrary(h); return; }

    unsigned char buffer[sizeof(unsigned int)];
    LONG status = p(NULL, buffer, sizeof(buffer), BCRYPT_USE_SYSTEM_PREFERRED_RNG);
    if (status == 0) { /* STATUS_SUCCESS == 0 */
        unsigned int seed;
        memcpy(&seed, buffer, sizeof(seed));
        srand(seed);
    }

    FreeLibrary(h);
}

const char *names[3] = {"Fido", "Spot", "Sparky"};

int dog_position() {
    return rand() % 99;
}

int dog_walking(int dog_position) {
    int time=rand()%10;
    int speed=rand()%10;
    int move=dog_position+(time*speed);
    printf("The dog is walking during %d seconds at a speed of %d m/s.\n", time, speed);
    return move;
}

int guess_position(int dog_new_position) {
    int guess;
    printf("Guess the new positionof the dog: ");
    scanf("%d", &guess);
    if (guess == dog_new_position) {
        printf("Congratulations! You guessed the correct position. \n");
        return 1;
    } else {
        printf("Incorrect guess! The correct position was %d. \n", dog_new_position);
        return 0;
    }
}

int main(void) {
    init_rng_with_bcrypt();
    for (int i = 0; i < 3; i++) {
        int pos = dog_position();
        printf("%s is at position %d\n", names[i], pos);
        int dog_new_position = dog_walking(pos);
        guess_position(dog_new_position);
    }
    return 0;
}